sys/linux/bpf_lsm.txt

```diff
@@ -0,0 +1,17 @@
+# Copyright 2020 syzkaller project authors. All rights reserved.
+# Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+include <uapi/linux/bpf.h>
+
+resource bpf_lsm_btf_id[bpf_btf_id]
+
+syz_btf_id_by_name$bpf_lsm(name ptr[in, string[bpf_lsm_func_names]]) bpf_lsm_btf_id (timeout[3000], prog_timeout[3000])
```

Comment:
Why do we need prog_timeout here?

Reply:
Do we plan to add any other specializations of this pseudo-syscall? If not, remove the `$bpf_lsm` suffix.

Reply:
It takes only about 40-50ms in my qemu VM:
```
# ./syz-execprog -debug -repeat=10 btf_id 2>&1 | grep syz_btf_id_by_name
#0 [1072ms] -> syz_btf_id_by_name$bpf_lsm(0x20000080)
#0 [1117ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [1374ms] -> syz_btf_id_by_name$bpf_lsm(0x20000200)
#0 [1375ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [1594ms] -> syz_btf_id_by_name$bpf_lsm(0x20000080)
#0 [1595ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#1 [1596ms] -> syz_btf_id_by_name$bpf_lsm(0x20000200)
#1 [1597ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [1620ms] -> syz_btf_id_by_name$bpf_lsm(0x20000080)
#0 [1690ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [2124ms] -> syz_btf_id_by_name$bpf_lsm(0x20000200)
#0 [2125ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [2445ms] -> syz_btf_id_by_name$bpf_lsm(0x20000080)
#0 [2446ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#1 [2447ms] -> syz_btf_id_by_name$bpf_lsm(0x20000200)
#1 [2448ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [2472ms] -> syz_btf_id_by_name$bpf_lsm(0x20000080)
#0 [2529ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [3084ms] -> syz_btf_id_by_name$bpf_lsm(0x20000200)
#0 [3086ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [3405ms] -> syz_btf_id_by_name$bpf_lsm(0x20000080)
#0 [3406ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#1 [3407ms] -> syz_btf_id_by_name$bpf_lsm(0x20000200)
#1 [3408ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [3432ms] -> syz_btf_id_by_name$bpf_lsm(0x20000080)
#0 [3482ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [3844ms] -> syz_btf_id_by_name$bpf_lsm(0x20000200)
#0 [3845ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [4165ms] -> syz_btf_id_by_name$bpf_lsm(0x20000080)
#0 [4166ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#1 [4169ms] -> syz_btf_id_by_name$bpf_lsm(0x20000200)
#1 [4171ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [4200ms] -> syz_btf_id_by_name$bpf_lsm(0x20000080)
#0 [4251ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [4574ms] -> syz_btf_id_by_name$bpf_lsm(0x20000200)
#0 [4575ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [4874ms] -> syz_btf_id_by_name$bpf_lsm(0x20000080)
#0 [4876ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#1 [4877ms] -> syz_btf_id_by_name$bpf_lsm(0x20000200)
#1 [4878ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [4906ms] -> syz_btf_id_by_name$bpf_lsm(0x20000080)
#0 [4960ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [5424ms] -> syz_btf_id_by_name$bpf_lsm(0x20000200)
#0 [5426ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [5645ms] -> syz_btf_id_by_name$bpf_lsm(0x20000080)
#0 [5646ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#1 [5647ms] -> syz_btf_id_by_name$bpf_lsm(0x20000200)
#1 [5648ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [5675ms] -> syz_btf_id_by_name$bpf_lsm(0x20000080)
#0 [5723ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [6134ms] -> syz_btf_id_by_name$bpf_lsm(0x20000200)
#0 [6135ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [6464ms] -> syz_btf_id_by_name$bpf_lsm(0x20000080)
#0 [6465ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#1 [6466ms] -> syz_btf_id_by_name$bpf_lsm(0x20000200)
#1 [6467ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [6488ms] -> syz_btf_id_by_name$bpf_lsm(0x20000080)
#0 [6543ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [6964ms] -> syz_btf_id_by_name$bpf_lsm(0x20000200)
#0 [6965ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [7174ms] -> syz_btf_id_by_name$bpf_lsm(0x20000080)
#0 [7175ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#1 [7176ms] -> syz_btf_id_by_name$bpf_lsm(0x20000200)
#1 [7177ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [7525ms] -> syz_btf_id_by_name$bpf_lsm(0x20000080)
#0 [7578ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [8144ms] -> syz_btf_id_by_name$bpf_lsm(0x20000200)
#0 [8145ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [8465ms] -> syz_btf_id_by_name$bpf_lsm(0x20000080)
#0 [8465ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#1 [8466ms] -> syz_btf_id_by_name$bpf_lsm(0x20000200)
#1 [8467ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [8708ms] -> syz_btf_id_by_name$bpf_lsm(0x20000080)
#0 [8751ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [9354ms] -> syz_btf_id_by_name$bpf_lsm(0x20000200)
#0 [9355ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#0 [9674ms] -> syz_btf_id_by_name$bpf_lsm(0x20000080)
#0 [9675ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
#1 [9677ms] -> syz_btf_id_by_name$bpf_lsm(0x20000200)
#1 [9678ms] <- syz_btf_id_by_name$bpf_lsm=0x705d errno=14 
```
3 seconds is a very long duration for a single test. Reduce it to what's really needed.
It probably should not really hang most of the time, so we can specify a timeout with some reserve, but 3000 still looks too much.

Reply:
> Do we plan to add any other specializations of this pseudo-syscall? If not, remove the `$bpf_lsm` suffix.

Currently we only input lsm hook names (those start with `bpf_lsm`) to the function. In the future one might use this pseudo-syscall to obtain the btf_id for other types of hooks.

Reply:
> It takes only about 40-50ms in my qemu VM:
> 
> 
> 3 seconds is a very long duration for a single test. Reduce it to what's really needed.
> It probably should not really hang most of the time, so we can specify a timeout with some reserve, but 3000 still looks too much.

It takes around 300ms to run in my VM so I will change it to 500ms (or should it be less?)

Reply:
> Currently we only pass lsm hook names (those start with `bpf_lsm`) to the function. In the future one might use this pseudo-syscall to obtain the btf_id for other types of hooks.

What can be done with other types of hooks? Since you already implemented syz_btf_id_by_name, maybe we now have some low hanging fruits that we can pick as well?

Reply:
> > It takes only about 40-50ms in my qemu VM:
> > 3 seconds is a very long duration for a single test. Reduce it to what's really needed.
> > It probably should not really hang most of the time, so we can specify a timeout with some reserve, but 3000 still looks too much.
> 
> It takes around 300ms to run in my VM so I changed it to 500ms (or should it be less?)

Let's keep 500, it sounds  reasonable.

Reply:
> > Currently we only pass lsm hook names (those start with `bpf_lsm`) to the function. In the future one might use this pseudo-syscall to obtain the btf_id for other types of hooks.
> 
> What can be done with other types of hooks? Since you already implemented syz_btf_id_by_name, maybe we now have some low hanging fruits that we can pick as well?

I will check what could be done.